cmake/git_version.cmake

```diff
@@ -0,0 +1,26 @@
+macro(set_git_version VERSION)
+  if(NOT Git_Found)
+    find_package(Git)
+  endif()
+
+  if(GIT_EXECUTABLE)
+    # Generate a git-describe version string from Git repository tags
+    execute_process(
+      COMMAND ${GIT_EXECUTABLE} describe --tags --dirty --match "v*"
+      WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+      OUTPUT_VARIABLE GIT_DESCRIBE_VERSION
+      RESULT_VARIABLE GIT_DESCRIBE_ERROR_CODE
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+      )
+    if(NOT GIT_DESCRIBE_ERROR_CODE)
+	    set(VERSION ${GIT_DESCRIBE_VERSION})
+    endif()
+  endif()
+
+  # Final fallback: Just use a bogus version string that is semantically older
+  # than anything else and spit out a warning to the developer.
+  if(NOT DEFINED VERSION)
+    set(VERSION v0.0.0-unknown)
+    message(WARNING "Failed to determine VERSION from Git tags. Using default version \"${VERSION}\".")
```

Comment:
Building without .git:
```output
CMake Warning at cmake/git_version.cmake:24 (message):
  Failed to determine VERSION from Git tags.  Using default version
  "VERSION".
Call Stack (most recent call first):
  CMakeLists.txt:35 (set_git_version)
```
I don't see why VERSION seems undefined.

Reply:
Ah, maybe when it is passed as a macro argument, it is defined as the value it already had in the caller context? Maybe we need
```diff
-  if(NOT DEFINED VERSION)
+  if(VERSION STREQUAL "")
```

Reply:
Nevermind. clearly it gets in the if branch.

Reply:
`VERSION` is somehow protected.
```diff
diff --git a/cmake/git_version.cmake b/cmake/git_version.cmake
index ef20bf3..935d2c9 100644
--- a/cmake/git_version.cmake
+++ b/cmake/git_version.cmake
@@ -22,5 +22,7 @@ macro(set_git_version VERSION)
   if(NOT DEFINED VERSION)
     set(VERSION v0.0.0-unknown)
     message(WARNING "Failed to determine VERSION from Git tags. Using default version \"${VERSION}\".")
+    set(_VERSION v0.0.0-unknown)
+    message(WARNING "Failed to determine VERSION from Git tags. Using default version \"${_VERSION}\".")
   endif()
 endmacro()
```

```output
EICrecon-git> CMake Warning at cmake/git_version.cmake:24 (message):
EICrecon-git>   Failed to determine VERSION from Git tags.  Using default version
EICrecon-git>   "VERSION".
EICrecon-git> Call Stack (most recent call first):
EICrecon-git>   CMakeLists.txt:35 (set_git_version)
EICrecon-git> 
EICrecon-git> CMake Warning at cmake/git_version.cmake:26 (message):
EICrecon-git>   Failed to determine VERSION from Git tags.  Using default version
EICrecon-git>   "v0.0.0-unknown".
EICrecon-git> Call Stack (most recent call first):
EICrecon-git>   CMakeLists.txt:35 (set_git_version)
EICrecon-git> 
```

Reply:
This is on cmake 3.24.1, btw